### PR TITLE
Pluggable cash selection using H2 as default implementation.

### PR DESCRIPTION
**Pull request by josecoll**
Enable pluggability of Cash Selection algorithms based on database provider.
Configure H2 as default implementation.

`CashSelection` implementations are loaded dynamically using a `ServiceLoader` from 
`META-INF/services/net.corda.contracts.asset.CashSelection`.
Default implementation is `net.corda.contracts.asset.cash.selection.CashSelectionH2Impl`

### CODE REVIEW

**Comment by josecoll** [comment]
Perhaps I should add some further documentation here:
https://docs.corda.net/writing-cordapps.html?highlight=plugins

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/Cash.kt [hunk 1]
```diff
@@ -45,6 +37,59 @@ import kotlin.concurrent.withLock
...
+ * Default implementation [CashSelectionH2Impl] uses H2 database and a custom function within H2 to perform aggregation.
+ * Custom implementations must implement this interface and declare their implementation in
+ * META-INF/services/net.corda.contracts.asset.CashSelection
+ */
+interface CashSelection {
+    companion object {
+        val instance = AtomicReference<CashSelection>()
+
+        fun getInstance(metadata: () -> java.sql.DatabaseMetaData): CashSelection {
+            if  (instance.get() == null) {
```
Comment (rick-r3) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/Cash.kt]: we could avoid double `get()` with `return instance.get() ?: { ... instance.set(...) ... return }`

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/Cash.kt [hunk 2]
```diff
@@ -45,6 +37,59 @@ import kotlin.concurrent.withLock
...
+        val instance = AtomicReference<CashSelection>()
+
+        fun getInstance(metadata: () -> java.sql.DatabaseMetaData): CashSelection {
+            return instance.get() ?: {
+                val _metadata = metadata()
+                val cashSelectionAlgos = ServiceLoader.load(CashSelection::class.java).toList()
+                val cashSelectionAlgo = cashSelectionAlgos.firstOrNull { it.isCompatible(_metadata) }
+                instance.set(cashSelectionAlgo ?: throw ClassNotFoundException("\nUnable to load compatible cash selection algorithm implementation for JDBC driver ($_metadata)." +
+                                                                               "\nPlease specify an implementation in META-INF/services/net.corda.finance.contracts.asset.CashSelection"))
+                instance.get()
```
Comment (rick-r3) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/Cash.kt]: Isn't this just `cashSelectionAlgo` ?

Comment (josecoll) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/Cash.kt]: Correct - and the thrown class should not be set in the atomic reference.

**Pull request merged**